### ISSUE TITLE
RFC: Potential redesign of how loaders/savers register with FileIO

### ISSUE THREAD

**Issue by KristofferC** [JuliaIO/FileIO.jl#189]
FileIO right now does quite a lot of "unorthodox" things which makes it a bit troublesome to upgrade to the new code loading, package management system. Examples of these are evaluating methods into other packages, reflecting on method tables at runtime etc.  In addition to being brittle, I believe these things heavily contribute to FileIO being slow to load and have quite complex code.

I suggest the following redesign for FileIO:

Packages that provide loading / saving functionality make themselves available to FileIO using something like:

```
add_loader(FileIO.Format{:PNG}, :ImageMagic, f_load)
add_loader_streaming(FileIO.Format{:PNG},  :ImageMagic, f_load_stream)
```

where `f_load` is a function defined in ImageMagic.jl that loads a file and `f_load_stream` is a function that loads from a buffer. Same applies for savers.

The crucial difference from before is that packages tell FileIO that they are available, FileIO itself doesn't try to find available loaders by looping through method tables and injecting load functions into packages etc.

Internally, FileIO would have a collection of loaders (and savers etc)

```jl
const loaders = Dict{FileIO.Format, Vector{Loader}}()
```

where, conceptually, a `Loader` is a

```jl
struct Loader
    package::Symbol
    load_function #::Function
end
```

The loader function `add_loader` would then look something like

```jl
function add_loader(t::Type{FileIO.Format{T}}, pkg::Symbol, f) where {T}
    haskey(loaders, t) || loaders[t] = Vector{Loader}()
    push!(loaders, Loader(pkg, f))
end
```

FileIO also has a registry which has the following functionalities

* From a file, return what file type the file is.
* From a file type, provide a list of possible packages that can load / save that file type.
* Also provide a priority list in case multiple loaders / savers are available.

This is already provided in `registry.jl` and it could likely be kept with few changes. The changes I can see is that the packages cannot be identified by just the name anymore so should use a UUID as well. 

Upon doing `FileIO.load("xxx.yyy")` the following steps are taken.

1. Use the registry to determine the file type (based on extension or magic bytes), say `FileIO.Format{T}`.
2. Based on the file type, look up pontential loaders in the `loaders` dictionary (`loaders[FileIO.Format{T}]`)
3. If there are multiple loaders, determine which one based on the priority list in the registry.
4. When a loader is determined, call the registered loading function for that loader.
5. If no loader was found, display an error message and give a suggestion of possible package that can load the file.

The following features are removed from FileIO:

* Automatic installation of packages and automatic loading of packages

Suggestion of packages to install / load can still be done by looking at the registry but the action to do so is left to the user / library.

cc @SimonDanisch, @davidanthoff, @fredrikekre

**Comment by SimonDanisch** [comment]
Sounds good to me!

**Comment by timholy** [comment]
It will be a bit of a loss to give up on automatic loading:
```julia
julia> img = load("photo.png");
ERROR: please import one of the following packages: ImageMagick, OpenImageIO, JuliaPNG, Fiona'sFastPNGLoader
Stacktrace:
 [1] top-level scope at none:0

julia> import Fiona'sFastPNGLoader

julia> img = load("photo.png");
```
Users may experience a bit of "but how do I choose?" angst. And code will get a bit uglier if, for example, ImageMagick is best on Linux but QuartzImageIO is best on OSX.

Nevertheless, I agree that's probably for the best.

**Comment by KristofferC** [comment]
1. and that it is much easier and better for performance for FileIO to reason about things in a `Dict` over things in a method table.

**Comment by davidanthoff** [comment]
I think that sounds like a great plan.

A few pain points that I observed over the years that we might as well clear out with this redesign as well:
- There is a fair bit of error catching and then trying the next thing going on, which I'm not a fan of. Here is an example: VegaLite.jl provides ``show`` methods for PDF, PNG, SVG etc. MIME types. For those files we have a writer registered in FileIO.jl that just uses these ``show`` methods to write things to disc. Sometimes though we have a plot that has an error, and in that case one obviously can't save the plot. But if I throw an error in the ``save`` function, FileIO.jl will just try the next PNG saver that it has, i.e. ImageMagic or something, and in the end the error message a user sees is really not helpful. I think we should redesign this so that a) errors are errors and are not caught and "hidden", and b) a writer should be able to signal to FileIO.jl without an exception that it can't handle a given object. Maybe just a return value or something. If it does have to be an exception, maybe we could define an exception type and FileIO.jl would only handle those exception types.
- On the writer side, I'm wondering in general whether a design that makes more use of ``show`` methods for the various MIME types might be less involved... We do sort of handle that for images right now by having a writer in FileIO.jl that is registered in the registry, but there might be a more general design along those lines.
- I think for loaders, at the end of the day it would be good if the user facing ``load`` function offers an option to a) manually specify what file type something is and b) overwrite the priority list in the registry for a given file type, i.e. also manually specify which loader to use if multiple are around.
- Here is another observation, but I'm not sure what to make of it, really: for the Queryverse.jl story, I'm generally return instances of types like ``CSVFile``, ``ExcelFile`` etc. from the ``load`` function. Those instances don't have the data loaded, they are more like lazy placeholders (i.e. they have a field with the filename). Only when you pass that value to a constructor for a type that can hold tabular data, is the data actually loaded. For example, ``a = load("foo.csv")`` doesn't load anything, only when one does ``DataFrame(a)`` is the stuff actually loaded from disc. That seems a very different pattern from what is done with images, right? Maybe it is just fine to have these two models, and things work, just wanted to mention it that we seem to have two different philosophies right now what ``load`` actually returns.

**Comment by oxinabox** [comment]
> Maybe just a return value or something. If it does have to be an exception, maybe we could define an exception type and FileIO.jl would only handle those exception types.

Maybe an `FeatureNotSupportedError`. which could take a message so the exact feature mentioned is covered. That way if everything fails  FileIO can report as such and explain why each did not.
I imagine the main reason for trying something else is that it is a feature of the format that this particular library doesn't support yet.

**Comment by timholy** [comment]
@oxinabox,
| Over simply overloading:
|`FileIO.load(imgfile::File{format"PNG"}) = f_load(imgfile)`

As you say, the problem with that is that then there can be only one loader for `format"PNG"`. Consequently we currently ask that [packages avoid overloading `FileIO.load`](https://github.com/JuliaIO/FileIO.jl#implementing-loaderssavers), and we have FileIO scope calls to `load` with the module name of the loader.

**Comment by tlnagy** [comment]
I was profiling some for `OMETIFF.jl` and I was dumbstruck at the size of FileIO.jl's overhead.

```julia
julia> using FileIO, OMETIFF, BenchmarkTools

julia> @benchmark load("big_image.ome.tif")
BenchmarkTools.Trial: 
  memory estimate:  3.52 GiB
  allocs estimate:  164487
  --------------
  minimum time:     11.844 s (1.98% GC)
  median time:      11.844 s (1.98% GC)
  mean time:        11.844 s (1.98% GC)
  maximum time:     11.844 s (1.98% GC)
  --------------
  samples:          1
  evals/sample:     1

julia> open("big_image.ome.tif") do f
           s = Stream(format"OMETIFF", f, OMETIFF.extract_filename(f))
           @benchmark img = OMETIFF.load($s)
       end
BenchmarkTools.Trial: 
  memory estimate:  3.52 GiB
  allocs estimate:  164159
  --------------
  minimum time:     4.029 s (3.84% GC)
  median time:      4.131 s (4.48% GC)
  mean time:        4.131 s (4.48% GC)
  maximum time:     4.233 s (5.09% GC)
  --------------
  samples:          2
  evals/sample:     1

```

Almost a 3x speedup by avoiding FileIO! Also, the overhead appears to be constant so even the most trivial operation takes 7s to pass through FileIO on my machine.

**Comment by timholy** [comment]
That's insane. FWIW other formats aren't remotely close to this. What's going on?

**Comment by tlnagy** [comment]
This appears to be a rogue header check function's fault, see #241 for details. I moved the discussion over there to not contaminate this discussion further since it's not the fault of the core FileIO code.

**Comment by NicholasWMRitchie** [comment]
I have a slightly different (and specialized use) for a FileIO-like functionality.  I'd like to be able to add FileIO-like file sniffing and automatic loading for some specialized (X-ray spectra) file types.  It doesn't really make sense to me to add them to the the FileIO registry as the audience for them is small.  However, I'd like to be able to register the loaders/savers when the user imports my library.  This doesn't seem to be possible with FileIO as it is currently designed. It seems that KristofferC's suggestion would facilitate my use but would also make FileIO less convenient for common file types like images.

**Comment by alhirzel** [comment]
@NicholasWMRitchie I have the same needs, and I think they can be met with `__init__`. I think documenting this as-is could go a long way toward capturing how FileIO is being used, and could inform a re-design. Is re-design still being considered?

**Comment by timholy** [comment]
I believe all this was fixed in #295. You register via `Base.PkgId` or the module directly if the module is already loaded.
